### ISSUE TITLE
Improvements in Logging

### ISSUE THREAD

**Issue by amshinde** [kata-containers/runtime#2650]
There are gaps in our logging infrastructure that need to be addressed. I am opening this issue to list down those gaps:

- [ ] We send our qemu logs to a file called qemu.log under /var/run/vc. However this file is always empty. It is essential that we capture logs from the hypervisor. We need to check and implement this for other hypervisors - firecracker and cloud-hypervisor as well.
- [ ] We can use a console to log in to the virtual machine. However when the console is used, logs from the agent and the VM are no longer sent to the journal. This could potentially be solved by having another console socket passed to the hypervisor.
- [ ] If the agent/VM/hypervisor crashes, all logs are potentially lost. We need to make sure that logs are available in all cases.

**Comment by amshinde** [comment]
cc @gnawux @jodh-intel @bergwolf @jcvenegas @devimc

**Comment by bporter816** [comment]
Hi! @tommyrousey, @maomarcus, and @yasv123 and I are students at UT Austin, and we're looking to make some open source contributions over the course of the rest of the year as a part of our virtualization class. Could we be assigned to this issue?

**Comment by devimc** [comment]
@bporter816 thanks, assigning this to you

**Comment by ariel-adam** [comment]
@bporter816 are you working on this issue or should we assign it to someone else?

**Comment by bporter816** [comment]
Yes, we plan to complete this soon.

**Comment by bporter816** [comment]
@amshinde @devimc We're gotten around to looking at this, and wanted to make sure we understand the relevant parts of the logging system to work on the issue. We're able to reproduce the `qemu.log` file being empty while running a container, and we're looking at the console socket at `console.sock` to try to understand how that works. One of the things we were wondering about is how to use a console to log into the virtual machine and generally how to interact with the console socket.

Would either of you or someone else be able to connect with us on Zoom sometime soon to walk us through this?
Thanks!

**Comment by devimc** [comment]
@bporter816 not too much to say about `qemu.log` and `console.sock`.  I have no idea why `qemu.log` is empty, I tried with --enable-debug but didn't work, @kata-containers/redhat any thoughts ?

Regarding to `console.sock`, the implementation of a new debug console shouldn't be a big deal, just add another virtconsole-socket pair (`-device virtconsole,chardev=charconsole1,id=console1 -chardev socket,id=charconsole1,path=debug-console.sock,server,nowait`) and somehow connect the vport in guest (/dev/vport1p0) with the bash process, take a look to the following chunk of code

https://github.com/kata-containers/agent/blob/25d74717af5a232fccb5f6e79df8f2e3a37be255/agent.go#L1424-L1433

**Comment by bporter816** [comment]
@amshinde @devimc We've dug into this some more and we believe that the `qemu.log` file is intended for severe QEMU issues such as virtual hardware errors. In most cases, this file is empty in our research. For example, we ran a plain `qemu` command with the `-D qemu.log` option outside of kata on a VM that successfully booted, and `qemu.log` was empty, even with all `-d` debug options enabled.

We were able to interact with the `qmp.sock` socket to capture events from the container such as when it was shutting down. This is a part of the QAPI provided by QEMU. These events also showed up in the standard `runtime.log` messages. Here's an example:
```
time="2020-11-30T22:41:55.88039594-06:00" level=info msg="{\"timestamp\": {\"seconds\": 1606797715, \"microseconds\": 880098}, \"event\": \"SHUTDOWN\", \"data\": {\"guest\": false, \"reason\": \"host-qmp-quit\"}}" arch=amd64 command=delete container=eb35ee2f446e9ea4bbec3555c354167c22be0d933c1bfb2526b01e87fda55031 name=kata-runtime pid=19678 sandbox=eb35ee2f446e9ea4bbec3555c354167c22be0d933c1bfb2526b01e87fda55031 source=virtcontainers subsystem=qmp
```

We were wondering what types of logs you are looking for from `qemu.log` - are they only meant to report low level QEMU errors, and should the file thus usually be empty?

**Comment by tommyrousey** [comment]
Hi @amshinde @devimc @bporter816,

I found an example of a `qemu.log` file in another ticket after reproducing the issue and finding the `qemu.log` empty with both Kata and a vanilla VM. This example found on issue https://github.com/kata-containers/runtime/issues/2978:
```
# cat qemu.log 
warning: Opening a block device as a file using the 'file' driver is deprecated
warning: Opening a block device as a file using the 'file' driver is deprecated
warning: Opening a block device as a file using the 'file' driver is deprecated
qemu-virtiofs-system-x86_64: VirtFS reply type 117 needs 14 bytes, buffer has 14, less than minimum
```
Also, from the QEMU documentation:
```
-D logfile
Output log in logfile instead of to stderr
```
I believe this supports @bporter816 's hypothesis that this is only for logging virtual hardware errors with the output usually going to stderr.

I also am also a bit confused on how the console was support to produce information similar to this. If this indeed not what is intended to be logged in `qemu.log`, what should then be logged using the console? Any clarification is helpful!

**Comment by devimc** [comment]
@tommyrousey @bporter816 

> I believe this supports @bporter816 's hypothesis that this is only for logging virtual hardware errors with the output usually going to stderr.

Ok, sounds like you got it, could you raise a pull request to fix/update the documentation?  thanks

https://github.com/kata-containers/runtime/blob/3f92a9c076f3e8e52ec1d54a0c7068f6c8d24e8a/cli/config/configuration-qemu.toml.in#L250-L255

https://github.com/kata-containers/runtime/blob/3f92a9c076f3e8e52ec1d54a0c7068f6c8d24e8a/cli/config/configuration-qemu-virtiofs.toml.in#L243-L248

**Comment by bporter816** [comment]
@amshinde We were wondering what sorts of logs are missing from the journal, in order to help us reproduce the second part of this issue. What's a good way to induce a log that would get captured by the console and not show up in the journal? And how would the introduction of a second console solve the issue? What does that change?

**Comment by amshinde** [comment]
Hi @bporter816, you can take a look at the logs expected to be generated by following this: 
https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#enable-full-debug

For example, after turning on debug for the agent, you may either see this log : "Serial channel type detected" or "Vsock channel type detected" in the journal. 
(Have a look at the agentLog calls in the agent to see other log statements. for eg : https://github.com/kata-containers/agent/blob/f9eab0fe9adb34e4f9f4a11f42a3eff983fd0659/channel.go#L102)

Once you start a debug console, you should still be able to see those logs in the journal. I hope it is clearer now.

**Comment by bporter816** [comment]
@amshinde @devimc @cmaf Looking for some context - we found the code in the runtime that sets up the connection with the agent via the proxy and shim:
https://github.com/kata-containers/runtime/blob/df055c8d506fe6e4d41a64d903d8c3fdf3d02592/virtcontainers/kata_agent.go#L743-L746 and https://github.com/kata-containers/runtime/blob/df055c8d506fe6e4d41a64d903d8c3fdf3d02592/virtcontainers/kata_agent.go#L1512-L1522

The logic for allowing messages to reach the system journal appears to be gated by checking if the debug console is enabled in the kernel params. Would it be sufficient to remove this part of these conditionals, or is there something else that requires the addition of a second console?

**Comment by devimc** [comment]
@bporter816 If you already implemented the function where the second  `console-socket` pair is attached to the hypervisor
> > We can use a console to log in to the virtual machine. However when the console is used, logs from the agent and the VM are no longer sent to the journal. This could potentially be solved by having another console socket passed to the hypervisor.

> > https://github.com/kata-containers/runtime/issues/2650#issuecomment-735894256

then yes, you should change that conditions

**Comment by bporter816** [comment]
@devimc We spent a while on this and weren't able to get it working so we'll leave it to someone with more expertise on the repo.
